<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU04_Optimización de vacunas</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 02. Zonas básicas de salud

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Encoding

Con la siguiente expresión se evitan problemas con el encoding al ejecutar el notebook. Es posible que deba ser eliminada o adaptada a la máquina en la que se ejecute el código.

In [ ]:
Sys.setlocale(category = "LC_ALL", locale = "es_ES.UTF-8")

[1] "es_ES.UTF-8/es_ES.UTF-8/es_ES.UTF-8/C/es_ES.UTF-8/C"

### Packages to use

- {readr} from _tidyverse_ for reading and writing csv files
- {tcltk} for selecting files and paths (if needed)
- {sf} for reading and writing files with spatial information 
- {dplyr} for data exploration

In [82]:
library(readr)
library(tcltk)
library(sf)
library(dplyr)


### Paths

In [83]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package (uncomment for using this option)

In [84]:
# file_data_01 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)
- Comentar si se usa la opción A

In [85]:
iFile <- "CU_04_05_01_zonasgeo.json"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_04_05_01_zonasgeo.json

#### Data file to dataframe

<font color='tomato'>Utilizamos en todos los notebooks SIEMPRE <b>data</b> como nombre de la matriz (dataframe) principal donde cargamos los datos. Si se importan más ficheros, se le pone un sufijo con número correlativo en dos cifras y una palabra informativa. Por ejemplo, data_01_poblacion. Se repiten las celdas de estructura y muestra de datos para dataframe.</font>

In [86]:
data <- st_read(file_data)

Reading layer `CU_04_05_01_zonasgeo' from data source 
  `/Users/emilio.lcano/academico/gh_repos/__transferencia/citizenlab/notebooks/II_data_processing/04_vacunas/Data/Input/CU_04_05_01_zonasgeo.json' 
  using driver `GeoJSON'
Simple feature collection with 286 features and 3 fields
Geometry type: MULTIPOLYGON
Dimension:     XY
Bounding box:  xmin: -4.579396 ymin: 39.8848 xmax: -3.052977 ymax: 41.16584
Geodetic CRS:  WGS 84


**NOTE**: This object is a special type of dataframe that includes the geometry of each observation. 

Estructura de los datos:

In [87]:
glimpse(data)

Rows: 286
Columns: 4
$ CODBDT    <int> 686213, 686214, 686215, 686216, 686217, 686218, 686219, 6862…
$ GEOCODIGO <chr> "001", "002", "003", "004", "005", "006", "007", "008", "009…
$ DESBDT    <chr> "Abrantes", "Acacias", "Adelfas", "Alameda", "Alameda de Osu…
$ geometry  <MULTIPOLYGON [°]> MULTIPOLYGON (((-3.718306 4..., MULTIPOLYGON ((…


Muestra de datos:

In [88]:
slice_head(data, n = 5)

CODBDT GEOCODIGO DESBDT           geometry                      
1 686213 001       Abrantes         MULTIPOLYGON (((-3.718306 4...
2 686214 002       Acacias          MULTIPOLYGON (((-3.707966 4...
3 686215 003       Adelfas          MULTIPOLYGON (((-3.666363 4...
4 686216 004       Alameda          MULTIPOLYGON (((-3.69947 40...
5 686217 005       Alameda de Osuna MULTIPOLYGON (((-3.561629 4...

## ETL Processes

### Transform data

* Para asegurar la trazabilidad, se guardan las transformaciones en un objeto diferente con el prefijo t
* Convertir los datos erróneos a válidos
* Calcular coordenadas del centroide de la zona sanitaria y guardar en columnas
* Calcular superficie de la zona para posteriormente calcular la densidad de población
* Eliminar información de área (polígonos)

#### Calcular área de la zona


In [89]:
sf_use_s2(FALSE)
tdata <- data |> 
    st_centroid()  
tdata <- tdata |>
    mutate(area = st_area(data)) |>
    bind_cols(st_coordinates(tdata)) |>
    st_drop_geometry()


Warning message in st_centroid.sf(data):
“st_centroid assumes attributes are constant over geometries of x”
Warning message in st_centroid.sfc(st_geometry(x), of_largest_polygon = of_largest_polygon):
“st_centroid does not give correct centroids for longitude/latitude data”


Muestra datos transformados:

In [90]:
tdata |> slice_head(n = 5)

CODBDT GEOCODIGO DESBDT           area             X         Y       
1 686213 001       Abrantes          1571618.8 [m^2] -3.726140 40.37898
2 686214 002       Acacias            771569.7 [m^2] -3.708702 40.40045
3 686215 003       Adelfas            854805.6 [m^2] -3.672841 40.40145
4 686216 004       Alameda            547596.1 [m^2] -3.697291 40.41036
5 686217 005       Alameda de Osuna 35137989.9 [m^2] -3.569774 40.47516

## Synthetic Data Generation

Estos datos no requieren tareas de este tipo.

## Fake Data Generation

Estos datos no requieren tareas de este tipo.

## Open Data

Los datos de origen fueron descargados de una fuente abierta, pero ya están preparados como fichero de entrada

## <font color='green'>Data Save</font>

<font color='tomato'> Identificamos los datos a guardar</font>

In [91]:
data_to_save <- tdata


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [92]:
caso <- "CU_04"
proceso <- '_05'
tarea <- "_02"
archivo <- ""
proper <- "_zonas"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario
* Descomentar si se usa esta opción

In [93]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(datos, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [94]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(oPath, file_save)
write_csv(tdata, path_out)

cat('File saved as: ')
path_out

File saved as: 

[1] "Data/Output/CU_04_05_02_zonas.csv"

#### Copia del fichero a Input

Será usado en otros notebooks

In [95]:
path_in <- paste0(iPath, file_save)
file.copy(path_out, path_in, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {readr} and {dplyr} packages must be installed. They are part of the _tidyverse_.
* The {sf} package must be installed. It needs some system requierements, check the package documentation at https://r-spatial.github.io/sf/
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path), as well as the input file

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* readr 2.1.3
* sf 1.0.9
* dplyr 1.0.10


### Data structures

- Los datos de origen fueron obtenidos de https://gestiona.comunidad.madrid/nomecalles/DescargaBDTCorte.icm#delimitacionesterritoriales y transformados previamente a un archivo .json, formato GeoJSON. Si la información geográfica cambia, se debería actualizar el fichero y volver a ejecutar todos los procesos.
- Los datos geográficos en el fichero JSON vienen con la proyección CRS WGS 84 (código 4326)
- Hay 286 zonas básicas de salud de las que se dispone de:
    * GEOCODIGO, string de tamaño tres con el código de la zona básica sanitaria que después se puede unir a otras tablas.
    * DESBDT, string de tamaño variable que contiene el nombre de la zona básica sanitaria y se usará para describir la zona.
    * CODBDT, entero que no utilizamos y no está claro para qué puede servir


## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

- Se han calculado las coordenadas de los centroides de las zonas de salud por si son útiles en futuros procesos

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben unir a estos datos los de indicadores del INE
- Se debe calcular la densidad de población de cada zona
- Los datos estarán relacionados con los de vacunación, meteorológicos y contaminación, de los que habrá varios registros por cada zona (uno por semana)
- También se relacionarán a través del municipio con los de tendencias en internet

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [96]:
# incluir código

## Consideraciones para despliegue en piloto

* El archivo de entrada se utilizará para la representación de mapas.

## Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados